### PR TITLE
[rollout, trainer] feat: ship routed_experts as ragged per-token payload

### PR DESCRIPTION
**Pull request by dengoswei**
### What does this PR do?

Ship rollout `routed_experts` on the driver as one numpy block per sequence in `non_tensor_batch`, sized to **attended tokens**, instead of a dense TensorDict tensor padded to the configured `prompt_length + response_length`. Worker unpad still accepts the previous dense tensor and rebuilds the same jagged NestedTensor. Positions the rollout never ran through MoE use a signed sentinel (`-1`) instead of expert 0.

This does **not** reintroduce int16 (`#7407`) or NestedTensor unpad (`#5240`); both are already on main. The leftover is the driver/wire representation: `_postprocess` still did `torch.cat` of ceiling-sized `[1, prompt+response, layers, topk]` tensors. Jagged NestedTensor cannot live in that TensorDict because `DataProto.concat` / `reorder` / `chunk` need a regular dim-0.

Related: `#5240`, `#7407`, `#6211` (transfer cost of padded routing), `#7158` / `#5519` (dtype/writability of the same payload).

### Checklist Before Starting

- [x] Search for similar PRs. Queries: [routed_experts](https://github.com/verl-project/verl/pulls?q=is%3Apr+routed_experts), [open routed_experts](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+routed_experts), [ragged routed](https://github.com/verl-project/verl/pulls?q=is%3Apr+ragged+routed_experts). Open hits (`#4517`, `#5506`) are AgentData / VL pad-size issues, not this transport. `#5240` already landed NestedTensor unpad; `#7407` already landed int16.
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

```bash
PYTHONPATH=. python -m pytest -q \
  tests/utils/test_routed_experts_ragged_on_cpu.py \
  tests/utils/test_padding_on_cpu.py \
  tests/trainer/ppo/test_rollout_moe_lb_metrics_on_cpu.py \
  tests/experimental/agent_loop/test_as_dict_routed_experts_sentinel_on_cpu.py
# 40 passed
```

Covered contracts:

- pack strips config padding and keeps the final-token unrecorded sentinel
- partial payloads and seq-length mismatches fail closed
- ragged → nested matches `unpad_input` offsets; row-count drift is rejected
- dense TensorDict path still unpads (adapter)
- `DataProto.concat` / `reorder` / `chunk` keep per-row blocks
- R2 (actor-recorded) vs R3 (rollout-recorded) conflict still fires when rollout routing lives in `non_tensor_batch`
- MoE load-balance metrics on ragged rows match the dense drop-last counts
- `AgentLoopOutput.as_dict` stays int16 and pads the tail with the sentinel, including read-only `np.frombuffer` input

No GPU/training run in this PR. The CPU tests exercise the actual pack/unpad helpers and DataProto ops, not mocked doubles of those helpers.

### API and Usage Example

No config or CLI change. Rollout routing replay (`enable_rollout_routing_replay=True`) keeps working; the payload is just smaller on the driver. Engine `prepare_model_inputs` still sees a jagged NestedTensor after `left_right_2_no_padding`.

Callers that read `batch.batch["routed_experts"]` on the driver should also look at `batch.non_tensor_batch["routed_experts"]`. `get_routed_experts_from_data_proto` does that.

### Design & Code Changes

- `AgentLoopWorker._postprocess` writes ragged rows into `non_tensor_batch` instead of `torch.cat` into the TensorDict.
- `_run_agent_loop` / `as_dict` fill unrecorded slots with `-1` (int16), not expert 0.
- `left_right_2_no_padding` converts ragged rows with the same `cu_seqlens` as `input_ids`; dense tensors keep the existing unpad path.
- Synthetic TransferQueue padding samples mirror the source representation (numpy row or tensor) and use the sentinel.
- R2/R3 exclusivity and MoE load-balance metrics see both representations.

Omega workspace sizing and Abbie-specific engine changes are intentionally out of this PR.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Per-file Ruff lint/format on the changed paths passed.
- [ ] Repository-wide pre-commit / CI — not run locally (no project venv in this worktree).
- [ ] Documentation — no public API/configuration change; behavior is internal to the routing payload.
- [x] CPU unit tests added; they are collected by the existing `cpu_unit_tests.yml` `test_*_on_cpu.py` glob.
- [ ] `ci-request` message — will send after CI is enabled on this PR.
- [x] Recipe submodule — not applicable.

### AI assistance disclosure

Grok assisted with porting, tests, and this description. The human submitter reviews every changed line before merge; agent review is not represented as human review.

### CODE REVIEW

**Comment by wuxibin89** [comment]
Please use V1 trainer, all fields including `routed_experts` are jagged tensor in TensorDict.

**Comment by dengoswei** [comment]
> Please use V1 trainer, all fields including `routed_experts` are jagged tensor in TensorDict.

Got. thx.. I wil revist the current verl roadmap..
